### PR TITLE
#2291  OAuth Integration (Github)

### PR DESCRIPTION
**Pull request by rohitjha941**
> What is Added
- OAuth (github) using Django Allauth & Django RestAuth 

>Testing Steps
- Create an OAuth APP on Github, Redirect URL should be `<frontend-url>/auth/oauth-redirect/github/`.
- Set the FrontEnd URL in the environment variable FRONTEND_URL.
- Change client id in frontend (login page)  
- Follow Django AllAuth Documentation to add a site 

>What is left?
- I need to set client id for frontend in  env, I am unable to find where env files of the frontend are stored

### CODE REVIEW

**Comment by RishabhJain2018** [comment]
Hi @rohitjha941, Can you please add the screencast of the feature.

**Comment by rohitjha941** [comment]
58746405-74fb9780-847b-11e9-980f-8c99a7f6f1cf

File: apps/accounts/views.py [hunk 1]
```diff
@@ -18,6 +18,19 @@
 from .permissions import HasVerifiedEmail
 
 
+from allauth.socialaccount.providers.github.views import GitHubOAuth2Adapter
+from allauth.socialaccount.providers.oauth2.client import OAuth2Client
+from rest_auth.registration.views import SocialLoginView
+from settings.common import FRONTEND_URL
```
Comment (deshraj) [inline apps/accounts/views.py]: Can you please change this line to `from django.conf import settings`

File: apps/accounts/views.py [hunk 2]
```diff
@@ -18,6 +18,19 @@
...
+
+
+class GithubLogin(SocialLoginView):
+    adapter_class = GitHubOAuth2Adapter
+    callback_url = FRONTEND_URL + "/auth/oauth-redirect/github/"
```
Comment (deshraj) [inline apps/accounts/views.py]: Change `FRONTEND_URL` to `settings.FRONTEND_URL`

File: apps/accounts/views.py [hunk 3]
```diff
@@ -49,3 +62,10 @@ def get_auth_token(request):
 
     response_data = {"token": "{}".format(token)}
     return Response(response_data, status=status.HTTP_200_OK)
+
```
Comment (deshraj) [inline apps/accounts/views.py]: Please remove all these extra lines.

File: evalai/urls.py [hunk 4]
```diff
@@ -27,6 +27,8 @@
 from drf_yasg import openapi
 
 from web import views
+from accounts import views as accountViews
+
```
Comment (deshraj) [inline evalai/urls.py]: Please remove this newline.

File: evalai/urls.py [hunk 5]
```diff
@@ -44,7 +46,8 @@
 )
 
 urlpatterns = [
-    url(r"^$", views.home, name="home"),
+    #url(r"^$", views.home, name="home"),
```
Comment (deshraj) [inline evalai/urls.py]: Please remove this commented line.

File: frontend/src/css/modules/auth.scss [hunk 6]
```diff
@@ -215,6 +215,52 @@ img.auth-logo {
     }
 }
 
+
```
Comment (deshraj) [inline frontend/src/css/modules/auth.scss]: Please remove line 218 and 219

File: frontend/src/css/modules/auth.scss [hunk 7]
```diff
@@ -215,6 +215,52 @@ img.auth-logo {
...
+        border:1px solid rgba(0,0,0,0.2);
+        padding:0.5em 1em;
+        img{
+            width:2em;
+            vertical-align: middle;
+            margin-right: 2px;
+        }
+    }
+}
+
```
Comment (deshraj) [inline frontend/src/css/modules/auth.scss]: Please remove line 262, 263

File: frontend/src/js/controllers/authCtrl.js [hunk 8]
```diff
@@ -285,8 +285,44 @@
             }
         };
 
+        vm.oauthLoginGithub = function(){
+            vm.startLoader("Logging you in");
+            var code = location.search;
+            code =  code.split('?code=')[1]
+            console.log(code)
```
Comment (deshraj) [inline frontend/src/js/controllers/authCtrl.js]: Please remove console statement.

File: frontend/src/js/controllers/authCtrl.js [hunk 9]
```diff
@@ -285,8 +285,44 @@
...
+
+            var parameters = {};
+            parameters.url = 'auth/oauth/github/';
+            parameters.method = 'POST';
+            parameters.data = {
+                "code": code,
+            };
+            console.log(parameters)
```
Comment (deshraj) [inline frontend/src/js/controllers/authCtrl.js]: Please remove console statement.

File: frontend/src/js/controllers/authCtrl.js [hunk 10]
```diff
@@ -285,8 +285,44 @@
...
+                    if (response.status == 200) {
+                        utilities.storeData('userKey', response.data.key);
+                        if ($rootScope.previousState) {
+                            $state.go($rootScope.previousState);
+                            vm.stopLoader();
+                        } else {
+                            $state.go('web.dashboard');
+                        }
+                    } else {
+                        alert("Something went wrong");
```
Comment (deshraj) [inline frontend/src/js/controllers/authCtrl.js]: Don't alert please. Redirect them to login page and show the message using toast.

File: frontend/src/js/controllers/authCtrl.js [hunk 11]
```diff
@@ -285,8 +285,44 @@
...
+                    }
+                },
+                onError: function(response){
+                    console.log(response)
```
Comment (deshraj) [inline frontend/src/js/controllers/authCtrl.js]: Please remove console statement. Redirect the user to login page and show the error message as toast.

File: frontend/src/js/controllers/authCtrl.js [hunk 12]
```diff
@@ -285,8 +285,44 @@
...
+                }
+            }
+            utilities.sendRequest(parameters, "no-header");
+
```
Comment (deshraj) [inline frontend/src/js/controllers/authCtrl.js]: Please remove the newline.

File: frontend/src/js/controllers/authCtrl.js [hunk 13]
```diff
@@ -285,8 +285,44 @@
...
+                    }
+                },
+                onError: function(response){
+                    console.log(response)
+                }
+            }
+            utilities.sendRequest(parameters, "no-header");
+            
+        };
+
```
Comment (deshraj) [inline frontend/src/js/controllers/authCtrl.js]: Please remove the newline on 322, 323

File: frontend/src/js/controllers/authCtrl.js [hunk 14]
```diff
@@ -285,8 +285,44 @@
...
+                }
+            }
+            utilities.sendRequest(parameters, "no-header");
+            
+        };
+        
+
         $rootScope.$on('$stateChangeStart', function() {
-            vm.resetForm();
+            vm.resetForm();
```
Comment (deshraj) [inline frontend/src/js/controllers/authCtrl.js]: Please revert this line. (there is extra space after `;`.)

File: frontend/src/views/web/analytics/host-analytics.html [hunk 15]
```diff
@@ -60,6 +60,7 @@ <h3 class="fs-20 rm-mt"><span>{{analytics.currentChallengeDetails.title}}</span>
                     <span class="show-count">
                         {{analytics.totalChallengeTeams}}
                     </span>
+                    <a class="participant-download-btn waves-effect waves-dark btn ev-btn-dark w-300 fs-12" ng-click="analytics.downloadChallengeParticipantTeams()">Download</a>
```
Comment (deshraj) [inline frontend/src/views/web/analytics/host-analytics.html]: Why is this line added in this PR? Please remove this.

File: frontend/src/views/web/auth/login.html [hunk 16]
```diff
@@ -20,8 +20,11 @@
         </div>
         <!-- password -->
         <div class="input-field align-left">
-            <input type="password" id="password" onpaste="return false;" class="dark-autofill" name="password" ng-model="auth.getUser.password" autocomplete="new-password" ng-minlength="8" required>
-            <span class="form-icon form-icon-dark"><i class="fa fa-key"></i></span>
+            <input type="{{canShowPassword ? 'text' : 'password'}}" id="password" onpaste="return false;" class="dark-autofill" name="password" ng-model="auth.getUser.password" autocomplete="new-password" ng-minlength="8" required>
```
Comment (deshraj) [inline frontend/src/views/web/auth/login.html]: Please revert this change since this is not related to oauth integration. Please feel free to open a new PR for this.

File: frontend/src/views/web/auth/login.html [hunk 17]
```diff
@@ -42,10 +45,24 @@
...
+                </a>
+        </div>
+ 
+
         <div>
             <p class="text-light-gray">
                 <span class="text-med-black">Start with a new account </span><a class="highlight-link" ui-sref="auth.signup" ng-click="auth.resetForm()"> Sign Up</a>
             </p>
         </div>
+
```
Comment (deshraj) [inline frontend/src/views/web/auth/login.html]: Please remove newline on 65 and 66

File: frontend/src/views/web/auth/oauth-redirect.html [hunk 18]
```diff
@@ -0,0 +1,4 @@
+<div ng-init="auth.oauthLogin()">
+
```
Comment (deshraj) [inline frontend/src/views/web/auth/oauth-redirect.html]: Please remove line 2,3

File: frontend/src/views/web/challenge/leaderboard.html [hunk 19]
```diff
@@ -122,7 +122,7 @@
                                                     id="baseline-badge" class="new badge orange-background"
                                                     data-badge-caption="B" ng-if="key.submission__is_baseline"></span>
                                             </td>
-                                            <td ng-repeat="score in key.result track by $index">{{score | number : 2}}
+                                            <td ng-repeat="(i, score) in key.result track by $index" >{{score | number : 2}}<span ng-if = "key.error != nil"> ± {{key.error[i] | number : 2}}</span>
```
Comment (deshraj) [inline frontend/src/views/web/challenge/leaderboard.html]: Can you please revert this change ? Not related to the oauth integration.

File: frontend/src/views/web/change-password.html [hunk 20]
```diff
@@ -6,17 +6,23 @@
                 <div class="pass-title">Change Password</div>
                 <form name="changepwdForm" ng-submit="changepwd.changePassword(changepwdForm.$valid)" novalidate>
                     <div class="input-field align-left">
-                        <input id="old_password" name="old_password" type="password" class="text-dark-black dark-autofill w-400" ng-model="changepwd.user.old_password" ng-minlength="8" required focus-if>
-                        <span class="form-icon form-icon-dark"><i class="fa fa-key"></i></span>
+                        <input id="old_password" name="old_password" type="{{canShowOldPassword ? 'text' : 'password'}}" class="text-dark-black dark-autofill w-400" ng-model="changepwd.user.old_password" ng-minlength="8" required focus-if>
```
Comment (deshraj) [inline frontend/src/views/web/change-password.html]: Please revert all changes in this file. Not related to oauth integration. Feel free to open a new PR.

File: frontend/src/views/web/featured-challenge/leaderboard.html [hunk 21]
```diff
@@ -51,7 +51,7 @@
                                         <tr ng-repeat="key in featured_challenge.leaderboard">
                                             <td class="align-left">{{$index+1}}</td>
                                             <td class="font-weight-b align-left">{{key.submission__participant_team__team_name}}</td>
-                                            <td class="align-right" ng-repeat="score in key.result track by $index">{{score | number : 2}}</td>
+                                            <td class="align-right" ng-repeat="(i, score) in key.result track by $index">{{score | number : 2}}<span ng-if = "key.error != nil">±{{key.error[i] | number : 2}}</span></td>
```
Comment (deshraj) [inline frontend/src/views/web/featured-challenge/leaderboard.html]: Please revert this.

File: settings/common.py [hunk 22]
```diff
@@ -179,13 +181,22 @@
...
 )
+SOCIALACCOUNT_PROVIDERS = {
+    'github': {
+        'SCOPE': [
+            'user',
+            'read:user'
+        ],
+    }
+}
+
```
Comment (deshraj) [inline settings/common.py]: Please remove line 198 and 199

File: settings/common.py [hunk 23]
```diff
@@ -288,6 +299,10 @@
 
 HOSTNAME = os.environ.get("HOSTNAME")
 
+
```
Comment (deshraj) [inline settings/common.py]: Please remove line 302 and 303

File: settings/common.py [hunk 24]
```diff
@@ -288,6 +299,10 @@
...
+
 SENDGRID_SETTINGS = {
     "TEMPLATES": {"CHALLENGE_INVITATION": "d-60825bcf014f4958bdb1b9173471d420"}
 }
+
+FRONTEND_URL =  os.environ.get("FRONTEND_URL" , "http://localhost:8888")
```
Comment (deshraj) [inline settings/common.py]: You might want to add this setting to

```
docker/prod/docker_staging.env
docker/prod/docker_production.env
docker/dev/docker.env

File: frontend/src/js/controllers/authCtrl.js (hunk 10 above)
Comment (rohitjha941) [reply frontend/src/js/controllers/authCtrl.js]: @deshraj  Never used Toast, Can you please share something regarding that

Comment (Sanji515) [reply frontend/src/js/controllers/authCtrl.js]: hey @rohitjha941 Show these messages over toast like here we are using https://github.com/Cloud-CV/EvalAI/blob/master/frontend/src/js/controllers/updateProfileCtrl.js#L51

I hope this will help!

Comment (rohitjha941) [reply frontend/src/js/controllers/authCtrl.js]: @Sanji515  Thank you

**Comment by RishabhJain2018** [comment]
Closing this one as #2412 is opened.
